File: Solutions/312. Burst Balloons/312. Burst Balloons.py
#   buttom up method
class Solution:
    def maxCoins(self, nums: List[int]) -> int:
        arr=[]
        arr.append(1)
        for num in nums:
            if num>0:
                arr.append(num)
        arr.append(1)
        n=len(arr)
        dp=[[0]*n for _ in range(n)]
        for i in range(n-1,-1,-1):
            for j in range(i+1,n,1):
                for k in range(i+1,j):
                    dp[i][j]=max(dp[i][j],dp[i][k]+dp[k][j]+arr[i]*arr[j]*arr[k])
        return dp[0][n-1]
        
        
        

        
#   top down recursive method with memoization, TLE in leetcode OJ (python3)
class Solution:
    def maxCoins(self, nums: List[int]) -> int:
        dp=[[0]*502 for _ in range(502)]
        def foo(arr,i,j):
            if(j-i<=1):
                return 0
            if not dp[i][j]:
                for k in range(i+1,j):
                    dp[i][j]=max(dp[i][j],arr[i]*arr[k]*arr[j]+foo(arr,i,k)+foo(arr,k,j))
            return dp[i][j]
        arr=[]
        arr.append(1)
        for num in nums:
            if num>0:
                arr.append(num)
        arr.append(1)
        return foo(arr,0,len(arr)-1)
            
            
            

File: Solutions/312. Burst Balloons/312. Burst Balloons.go
//buttom up method
func maxCoins(nums []int) int {
    var arr []int
    arr=append(arr,1)
    for _,num:=range(nums){
        if num>0{
            arr=append(arr,num)
        }
    }
    arr=append(arr,1)
    dp:=make([][]int,len(arr))
    for i,_:=range(dp){dp[i]=make([]int,len(arr))}
    for i:=len(arr)-1;i>=0;i-=1{
        for j:=i+1;j<len(arr);j+=1{
            for k:=i+1;k<j;k+=1{
                dp[i][j]=max(dp[i][j],dp[i][k]+dp[k][j]+arr[i]*arr[j]*arr[k])
            }
        }
    }
    return dp[0][len(arr)-1]
}

func max(i,j int)int{
    if(i>j){
        return i
    }
    return j
}

/*******************************************************************************/
//top down recursive method with memoization

func maxCoins(nums []int) int {
    dp:=make([][]int,502)
    for i:=0;i<502;i+=1{
        dp[i]=make([]int,502)
    }
    var arr []int
    arr=append(arr,1)
    for _,num:=range(nums){
        if num>0{
            arr=append(arr,num)
        }
    }
    arr=append(arr,1)
    return foo(&arr,&dp,0,len(arr)-1)
}

func foo(arr *[]int,dp *[][]int,i,j int)int{
    if j-i<=1{
        return 0
    }
    if (*dp)[i][j]==0{
        for k:=i+1;k<j;k+=1{
            (*dp)[i][j]=max((*dp)[i][j],foo(arr,dp,i,k)+foo(arr,dp,k,j)+(*arr)[i]*(*arr)[j]*(*arr)[k])
        }
    }
    return (*dp)[i][j]
}

func max(i,j int)int{
    if i<j{
        return j
    }
    return i
}